File: 0543_diameter-of-binary-tree/solution.go
package _543_diameter_of_binary_tree

type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}

func diameterOfBinaryTree(root *TreeNode) int {
	ans:=0
	var depth func(node *TreeNode) int
	depth = func(node *TreeNode) int {
		if node  == nil {
			return 0
		}
		l := depth(node.Left)
		r := depth(node.Right)
		ans = max(ans, l+r+1)
		return max(l, r) + 1
	}

	depth(root)
	return ans-1
}


func max( a, b int) int{
	if a > b{
		return a
	}
	return b
}

File: 0543_diameter-of-binary-tree/solution.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# https://leetcode-cn.com/problems/diameter-of-binary-tree/solution/er-cha-shu-de-zhi-jing-by-leetcode-solution/
class Solution:
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        ans = 0

        def depth(node):
            nonlocal ans
            if node is None:
                return 0
            left = depth(node.left)
            right = depth(node.right)
            ans = max(ans, left + right + 1)
            return max(left, right) + 1

        depth(root)
        return ans - 1
